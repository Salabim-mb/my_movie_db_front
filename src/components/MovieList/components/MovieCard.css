.MovieCard {
    text-align: justify;
    margin-left: 3%;
}

.ThumbnailImage {
    max-width: 180px;
}