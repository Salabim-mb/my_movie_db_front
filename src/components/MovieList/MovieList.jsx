import React, {useEffect, useState} from "react";
import Page from "../Page";
import MovieCard from "./components/MovieCard";
import {Alert, Spinner} from "react-bootstrap";
import {backend} from "../../constants/backend";

const getMovies = async () => {
    // jeszcze nie wiemy jakie będą headery do zapytania, ale piszę na czuja
    const headers = {
        "Access-Control-Allow-Origin": "*",
        "Content-Type": "application/json", // będziemy wysyłać dane jsonem
    };
    // zmienna domain jest zła, tu normalnie będzie url, ale trzymamy go w pliku .env
    // o .env nie chce mi się pisać za bardzo, wytłuamczę na żywo że to są
    // zmienne środowiskowe
    const url = backend.MOVIES;
    // bez await dostaniemy Promise, czyli obietnicę że to zostanie wypełnione
    // nie chcemy działać na promisach, więc dajemy await żeby poczekać aż promise zostanie
    // rozwiązany, po czym działamy na pierwszej "warstwie" odpowiedzi, czyli
    // metadanych (status HTTP itd) oraz body jako kolejny promise
    let res = await fetch(url, {
        headers,
        method: "GET"
        // do GETa nie potrzeba body, ale przy innych metodach czasami tak. W związku z tym wrzucamy wtedy
        // body: JSON.stringify( obiekt_przekazany_jako_argument )
        // powyższa metoda zamienia obiekt na jsona
    });

    if (res.status === 200) { // OK
        return await res.json(); // return await oznacza, że body naszej odpowiedzi będzie zparsowane
    } else {
        throw res.status; // jeżeli status nie jest ok to go łapiemy w bloku try/catch metody loadMovieList()
    }
};

const MovieList = () => {
   /*
       poniżej pokazany jest sposób zapisu stanu komponenetu funkcyjnego, taka składnia zastępuje konstruktor
       komponent klasowy (taki sam) umieszczę w MovieListClassComponent.jsx
       loading, error - boolean (stan początkowy false)
       movieList - lista, dlatego mamy useState([])
    */
    const [loading, setLoading] = useState(false);
    const [error, setError] = useState(false);
    const [movieList, setMovieList] = useState([]);

    useEffect(() => {
        // napisanie wewnętrznej metody pozwoli na użycie asynchroniczności
        // loadMovieList jest naszym wrapperem do fetcha
        const loadMovieList = async () => {
            setLoading(true);
            try {
                // czekamy se na odpowiedź z GET'a, ewentualnie na throw statusu HTTP
                // (najczęściej będzie to 400 Bad Request xd)
                let res = await getMovies();
                // mamy odpowiedź, teraz trzeba ją zmapować, bo backend może używać innych nazw pól

                setMovieList( res );

            } catch(e) {
                // nie wyszło z GET'em, więc można wypisać na konsolę to co rzuciliśmy w getMovies()
                // i ustawić flagę error na true
                console.log(e);
                setError(true);
            } finally {
                // niezależnie od wyniku GET'a, kończymy ładowanie
                setLoading(false);
            }
        };
        // wywołujemy funkcję, co prawda jest asynchroniczna, ale można to pominąć, bo mamy oczekiwanie w środku
        loadMovieList();
    }, []);

    // metoda na pokazanie alertu kiedy ustawi się error = true
    const renderErrorAlert = () => {
        return <Alert variant="danger">Something went wrong while trying to get movie list.</Alert>
    };

    // metoda na pokazanie spinnera ładowania kiedy loading = true
    const renderSpinner = () => {
        return <Spinner animation="border" role="status" />
    };

    const refreshMovieList = (id) => {
        let movies = movieList.filter(item => item.id !== id);
        setMovieList(movies)
    }

    return (
        <Page>
            {/*{ można wyszukiwarkę dopisać, czysto frontendowa }*/}
            {/*{ tutaj będzie lista, czyli zmapowana movieList, ale najpierw error i loading }*/}
            {
                error ? renderErrorAlert() :
                loading ? renderSpinner() :
                movieList.map((item) => (
                    <MovieCard key={item?.id} movie={item} setlist={refreshMovieList}/>
                ))
            }
        </Page>
    )
};

export default MovieList;