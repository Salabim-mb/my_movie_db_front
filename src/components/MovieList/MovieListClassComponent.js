import React from "react";
import MovieCard from "./components/MovieCard";
import Page from "../Page";
import {Alert, Spinner} from "react-bootstrap";
import {backend} from "../../constants/backend";

class MovieListClassComponent extends React.Component {

    constructor(props){
        super(props);
        this.state = {
            loading: false,
            error: false,
            movieList: []
        }
    };

    componentDidMount() {
        const loadMovieList = async () => {
            this.setState({loading: true});
            try {
                // czekamy se na odpowiedź z GET'a, ewentualnie na throw statusu HTTP
                // (najczęściej będzie to 400 Bad Request xd)
                let res = await this.getMovies();
                // mamy odpowiedź, teraz trzeba ją zmapować, bo backend może używać innych nazw pól
                this.setState({ movieList: this.mapMovies(res) });
            } catch(e) {
                // nie wyszło z GET'em, więc można wypisać na konsolę to co rzuciliśmy w getMovies()
                // i ustawić flagę error na true
                console.log(e);
                this.setState({error: true});
            } finally {
                // niezależnie od wyniku GET'a, kończymy ładowanie
                this.setState({loading: false});
            }
        };
        // wywołujemy funkcję, co prawda jest asynchroniczna, ale można to pominąć, bo mamy oczekiwanie w środku
        loadMovieList();
    }

    mapMovies = (data) => data.map({
        //... pola do używania w movieListm o ile używane są inne niż te z odpowiedzi od backendu
    });

    getMovies = async() => {
        let url = backend.MOVIES;
        let headers = {
            "Content-Type": "application/json"
        };

        let res = await fetch(url, {
            headers,
            method: "GET"
        });

        if (res.status === 200) {
            return await res.json();
        } else {
            throw res.status;
        }


    };

    renderErrorAlert = () => {
        return <Alert variant="danger">Something went wrong while trying to get movie list.</Alert>
    };

    renderSpinner = () => {
        return <Spinner animation="border" role="status" />
    };

    render() {
        let {loading, error, movieList} = this.state;
        let {renderErrorAlert, renderSpinner} = this;

        return (
            <Page>
                {/*{ można wyszukiwarkę dopisać, czysto frontendowa }*/}
                {/*{ tutaj będzie lista, czyli zmapowana movieList, ale najpierw error i loading }*/}
                {
                    error ? renderErrorAlert() :
                        loading ? renderSpinner() :
                            movieList.map((item) => (
                                <MovieCard key={item.id} movie={item} />
                            ))
                }
            </Page>
        )
    }
};

export default MovieListClassComponent;