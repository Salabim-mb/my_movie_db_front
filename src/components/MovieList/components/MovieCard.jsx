import React from "react";
import Card from "react-bootstrap/Card";
import Button from "react-bootstrap/Button";
import './MovieCard.css';

const MovieCard = ({movie}) => (
    <div className="card mb-3">
        <div className="row no-gutters">
            <div className="ThumbnailImage">
                <Card.Img src={movie.movieImage}/> {/*{movie.movieImage}*/}
            </div>
            <div className="col-sm-9">
                <div className="MovieCard">
                    <div className="card-body">
                        <Card.Title>
                            {movie.movieName}
                        </Card.Title>
                        <Card.Subtitle className="mb-2 text-muted">
                            {movie.movieDirector}
                        </Card.Subtitle>
                        <p className="card-text"><small className="text-muted">{movie.movieGenre}</small></p>
                        <Card.Text>
                            {movie.movieReleaseYear}
                        </Card.Text>
                        <hr/>
                        <p className="card-text">{movie.movieDescription} </p>
                    </div>
                </div>
            </div>
        </div>
    </div>

);

export default MovieCard;