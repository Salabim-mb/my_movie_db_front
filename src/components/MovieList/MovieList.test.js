import React from "react";
import MovieList from "./MovieList";
import MovieCard from "./components/MovieCard";
import {MemoryRouter} from "react-router";
import {render, waitForElement} from '@testing-library/react';

jest.mock("./components/MovieCard");

describe("MovieList", () => {
    let apiFail;
    let res;

    beforeAll(() => {
        global.fetch = jest.fn().mockImplementation((input, init) => {
            return new Promise((resolve) => {
                if (apiFail) {
                    resolve({status: 500});
                } else {
                    resolve({status: 200, json: () => Promise.resolve(res)});
                }
            });
        });
    });

    beforeEach(() => {
        jest.clearAllMocks();
        MovieCard.mockImplementation((props) => <div>{props?.movie?.title}</div>);
        apiFail = false;
        res = [
            {
                id: 1,
                title: "Titanic"
            }, {
                id: 2,
                title: "Batman"
            }
        ]
    });

    it("should match snapshot", async () => {
        const {container, getByText} = render(
            <MemoryRouter>
                <MovieList />
            </MemoryRouter>
        );

        await waitForElement(() => getByText("Titanic"));

        expect(container).toMatchSnapshot();
    });

    it("should fetch correct data", async () => {
        const {getByText, queryByText} = render(
            <MemoryRouter>
                <MovieList />
            </MemoryRouter>
        );
        await waitForElement(() => getByText("Titanic"));

        expect(getByText("Titanic")).toBeInTheDocument();
        expect(queryByText("Something went wrong", {exact: false})).toBe(null);
        //expect(queryByText("Something went wrong", {exact: false})).not.toBeInTheDocument();
    });

    it ("should render error alert", async () => {
        apiFail = true;

        const {getByText, queryByText} = render(
            <MemoryRouter>
                <MovieList />
            </MemoryRouter>
        );

        await waitForElement(() => getByText("Something went wrong", {exact: false}));

        expect(getByText("Something went wrong", {exact: false})).toBeInTheDocument();
        expect(queryByText("Batman")).not.toBeInTheDocument();
    });
});