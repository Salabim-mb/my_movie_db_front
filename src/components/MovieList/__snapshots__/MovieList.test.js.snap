// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList should match snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="card"
    >
      <div
        class="card-body"
      >
        <div>
          Titanic
        </div>
        <div>
          Batman
        </div>
      </div>
    </div>
  </div>
</div>
`;
